File: prototype/rails/src/app/actions/orders/place_order_test.rb
require "commerce_test_case"

module Orders
  class PlaceOrderTest < CommerceTestCase
    def test_it_turns_the_cart_into_an_order_the_customer_can_pay_for
      order = place(customer)

      assert_equal Domain::Orders::OrderStatus::AWAITING_PAYMENT, order.status
      assert_equal 45_000, order.subtotal_cents
      assert_equal 45_000, order.total_cents
      assert_equal moment("2026-08-20 09:00:00"), order.placed_at
      assert_nil order.finalized_at
    end

    def test_a_guest_places_an_order_that_waits_for_verification
      assert_equal Domain::Orders::OrderStatus::PENDING_VERIFICATION, place(anonymous_customer).status
    end

    def test_it_copies_the_shipping_address_onto_the_order
      order = place(customer)

      assert_equal "Ada Lovelace", order.shipping_name
      assert_equal "12 Analytical Way", order.shipping_line1
      assert_nil order.shipping_line2
      assert_equal "EC1A 1BB", order.shipping_postal_code
      assert_equal "GB", order.shipping_country
    end

    def test_it_snapshots_the_title_and_price_of_every_item
      art = listing(seller, title: "Harbour at Dusk", price_cents: 45_000)

      item = order_for(customer, art).items.sole

      assert_equal "Harbour at Dusk", item.title
      assert_equal 45_000, item.unit_price_cents
      assert_equal art.seller_id, item.seller_id
    end

    def test_it_splits_the_order_into_one_fulfillment_per_seller
      painting = listing(seller("Blue Kiln Studio"), price_cents: 45_000)
      print = listing(seller("Rye Press"), price_cents: 10_000)

      order = order_for(customer, painting, print)

      assert_equal 55_000, order.subtotal_cents
      assert_equal(
        [[painting.seller_id, 45_000, 4500, 40_500], [print.seller_id, 10_000, 1000, 9000]],
        order.fulfillments.order(:seller_id).map { |f| [f.seller_id, f.subtotal_cents, f.fee_cents, f.net_cents] }
      )
    end

    def test_every_fulfillment_starts_awaiting_shipment
      status = place(customer).fulfillments.sole.status

      assert_equal Domain::Orders::FulfillmentStatus::AWAITING_SHIPMENT, status
    end

    def test_it_takes_the_stock_the_order_claims
      art = listing(seller, quantity: 3)
      cart = cart_for(customer)
      Carts::AddToCart.new.call(cart: cart, listing: art, quantity: 2, now: moment("2026-08-20 08:00:00"))

      PlaceOrder.new.call(cart: cart, purchaser: purchaser(cart.customer), shipping: shipping_address,
                          now: moment("2026-08-20 09:00:00"))

      art.reload
      assert_equal 1, art.quantity
      assert_equal Domain::Listings::ListingStatus::FOR_SALE, art.status
    end

    def test_the_last_of_a_listing_marks_it_sold
      art = listing(seller, quantity: 1)

      order_for(customer, art)

      art.reload
      assert_equal 0, art.quantity
      assert_equal Domain::Listings::ListingStatus::SOLD, art.status
    end

    def test_it_empties_the_cart
      buyer = customer
      cart = cart_holding(buyer, listing(seller))

      PlaceOrder.new.call(cart: cart, purchaser: purchaser(buyer), shipping: shipping_address,
                          now: moment("2026-08-20 09:00:00"))

      assert_empty cart.reload.items
    end

    def test_it_refuses_an_empty_cart
      buyer = customer

      assert_raises(ArgumentError) do
        PlaceOrder.new.call(cart: cart_for(buyer), purchaser: purchaser(buyer), shipping: shipping_address,
                            now: moment("2026-08-20 09:00:00"))
      end
    end

    private

    def place(buyer)
      order_for(buyer, listing(seller, price_cents: 45_000))
    end
  end
end

File: prototype/node/src/app/actions/orders/place-order.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { finalizeOrder } from './finalize-order.ts'
import { placeOrder } from './place-order.ts'
import { addToCart } from '../carts/add-to-cart.ts'
import { currentCart } from '../carts/current-cart.ts'
import { runInTransaction } from '../transaction.ts'
import type { ActionContext } from '../action-context.ts'
import type { AppDatabase } from '../../db/database.ts'
import { toTimestamp } from '../../db/timestamp.ts'
import {
  APPROVED_CARD,
  SHIPPING_ADDRESS,
  cartHolding,
  createAdmin,
  createCustomer,
  createListing,
  createSeller,
  openCommerceWorld,
  placedOrder,
} from '../../test/commerce-world.ts'

test('it turns the cart into an order a verified customer can pay for', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop, { priceCents: 45_000 })

  const order = await placedOrder(context, buyer, [art.id])

  assert.equal(order.status, 'awaiting_payment')
  assert.equal(order.subtotalCents, 45_000)
  assert.equal(order.totalCents, 45_000)
  assert.equal(order.placedAt, new Date('2026-08-20T09:00:00.000Z').toISOString())
  assert.equal(order.finalizedAt, null)
})

test('a guest places an order that waits for verification', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const guest = await createCustomer(context, { isVerified: false })
  const art = await createListing(context, shop)

  const order = await placedOrder(context, guest, [art.id], { isVerified: false })

  assert.equal(order.status, 'pending_verification')
})

test('it copies the shipping address onto the order', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop)

  const order = await placedOrder(context, buyer, [art.id])

  assert.equal(order.shippingName, SHIPPING_ADDRESS.name)
  assert.equal(order.shippingLine1, SHIPPING_ADDRESS.line1)
  assert.equal(order.shippingLine2, null)
  assert.equal(order.shippingCity, SHIPPING_ADDRESS.city)
  assert.equal(order.shippingRegion, SHIPPING_ADDRESS.region)
  assert.equal(order.shippingPostalCode, SHIPPING_ADDRESS.postalCode)
  assert.equal(order.shippingCountry, SHIPPING_ADDRESS.country)
})

test('it snapshots the title, unit price, and seller of every item', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop, { title: 'Harbour at Dusk', priceCents: 45_000 })

  const order = await placedOrder(context, buyer, [art.id])
  const [item] = await readOrderItems(world.db, order.id)

  assert.equal(item?.title, 'Harbour at Dusk')
  assert.equal(item?.unitPriceCents, 45_000)
  assert.equal(item?.sellerId, shop)
})

test('it splits the order into one fulfillment per seller with a 10% fee', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const painter = await createSeller(context, 'Blue Kiln Studio')
  const printer = await createSeller(context, 'Rye Press')
  const buyer = await createCustomer(context)
  const painting = await createListing(context, painter, { priceCents: 45_000 })
  const print = await createListing(context, printer, { priceCents: 10_000 })

  const order = await placedOrder(context, buyer, [painting.id, print.id])

  assert.equal(order.subtotalCents, 55_000)
  assert.deepEqual(await readFulfillments(world.db, order.id), [
    { sellerId: painting.sellerId, status: 'awaiting_shipment', subtotalCents: 45_000, feeCents: 4_500, netCents: 40_500 },
    { sellerId: print.sellerId, status: 'awaiting_shipment', subtotalCents: 10_000, feeCents: 1_000, netCents: 9_000 },
  ])
})

test('it takes the stock the order claims', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop, { quantity: 3 })

  const cart = await currentCart(context, buyer)
  await addToCart(context, { cartId: cart.id, listingId: art.id, quantity: 2 })
  await placeOrder(context, {
    cartId: cart.id,
    purchaser: { id: buyer, email: 'ada@example.test', isEmailVerified: true },
    shipping: SHIPPING_ADDRESS,
  })

  assert.deepEqual(await readStock(world.db, art.id), { quantity: 1, status: 'for_sale' })
})

test('the last of a listing marks it sold', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop, { quantity: 1 })

  await placedOrder(context, buyer, [art.id])

  assert.deepEqual(await readStock(world.db, art.id), { quantity: 0, status: 'sold' })
})

test('it empties the cart', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop)

  const order = await placedOrder(context, buyer, [art.id])
  const cart = await currentCart(context, buyer)

  assert.equal(order.customerId, buyer)
  assert.deepEqual(await readCartItems(world.db, cart.id), [])
})

test('it refuses an empty cart', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const buyer = await createCustomer(context)
  const cart = await currentCart(context, buyer)

  await assert.rejects(
    () =>
      placeOrder(context, {
        cartId: cart.id,
        purchaser: { id: buyer, email: 'ada@example.test', isEmailVerified: true },
        shipping: SHIPPING_ADDRESS,
      }),
    RangeError,
  )
})

test('it refuses a cart holding a listing an admin removed and places nothing', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop, { title: 'Harbour at dusk' })
  const cartId = await cartHolding(context, buyer, [art.id])
  await removeListing(context, art.id, await createAdmin(context))

  const placement = await placeOrder(context, {
    cartId,
    purchaser: { id: buyer, email: 'ada@example.test', isEmailVerified: true },
    shipping: SHIPPING_ADDRESS,
  })

  assert.deepEqual(placement, {
    ok: false,
    unavailable: [{ listingId: art.id, title: 'Harbour at dusk', reason: 'removed' }],
  })
  assert.equal((await readOrders(world.db)).length, 0)
  assert.equal((await readCartItems(world.db, cartId)).length, 1)
})

test('it refuses a cart holding a listing another buyer already took', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const quicker = await createCustomer(context)
  const art = await createListing(context, shop, { title: 'Harbour at dusk', quantity: 1 })
  const cartId = await cartHolding(context, buyer, [art.id])
  await placedOrder(context, quicker, [art.id])

  const placement = await placeOrder(context, {
    cartId,
    purchaser: { id: buyer, email: 'ada@example.test', isEmailVerified: true },
    shipping: SHIPPING_ADDRESS,
  })

  assert.deepEqual(placement, {
    ok: false,
    unavailable: [{ listingId: art.id, title: 'Harbour at dusk', reason: 'sold_out' }],
  })
  assert.equal((await readOrders(world.db)).length, 1)
})

test('placement and the charge roll back together', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const shop = await createSeller(context)
  const buyer = await createCustomer(context)
  const art = await createListing(context, shop)
  const cartId = await cartHolding(context, buyer, [art.id])

  await assert.rejects(() =>
    runInTransaction(context, async (transacted) => {
      const placement = await placeOrder(transacted, {
        cartId,
        purchaser: { id: buyer, email: 'ada@example.test', isEmailVerified: true },
        shipping: SHIPPING_ADDRESS,
      })
      assert.equal(placement.ok, true)
      await finalizeOrder(transacted, { orderId: placement.order.id, cardNumber: APPROVED_CARD })

      throw new Error('the request failed after the charge')
    }),
  )

  assert.equal((await readOrders(world.db)).length, 0)
  assert.equal((await readPayments(world.db)).length, 0)
  assert.equal((await readCartItems(world.db, cartId)).length, 1)
})

async function removeListing(
  { db, clock }: ActionContext,
  listingId: number,
  adminId: number,
): Promise<void> {
  await db
    .insertInto('listingRemovals')
    .values({
      listingId,
      adminId,
      kind: 'temporary',
      reason: 'Reported as a reproduction.',
      createdAt: toTimestamp(clock.now()),
      liftedAt: null,
    })
    .execute()
}

async function readOrders(db: AppDatabase) {
  return db.selectFrom('orders').select('id').execute()
}

async function readPayments(db: AppDatabase) {
  return db.selectFrom('payments').select('id').execute()
}

async function readOrderItems(db: AppDatabase, orderId: number) {
  return db
    .selectFrom('orderItems')
    .select(['title', 'unitPriceCents', 'sellerId'])
    .where('orderId', '=', orderId)
    .execute()
}

async function readFulfillments(db: AppDatabase, orderId: number) {
  return db
    .selectFrom('fulfillments')
    .select(['sellerId', 'status', 'subtotalCents', 'feeCents', 'netCents'])
    .where('orderId', '=', orderId)
    .orderBy('sellerId')
    .execute()
}

async function readStock(db: AppDatabase, listingId: number) {
  return db
    .selectFrom('listings')
    .select(['quantity', 'status'])
    .where('id', '=', listingId)
    .executeTakeFirstOrThrow()
}

async function readCartItems(db: AppDatabase, cartId: number) {
  return db.selectFrom('cartItems').select('id').where('cartId', '=', cartId).execute()
}
